nvme.c

```diff
@@ -1893,6 +1893,210 @@ static int get_supp_cap_config_log(int argc, char **argv, struct command *cmd,
 	return err;
 }
 
+static int io_mgmt_send(int argc, char **argv, struct command *cmd, struct plugin *plugin)
+{
+	const char *desc = "I/O Management Send";
+	const char *mos = "Management Operation Specific";
+	const char *mo = "Management Operation";
+	const char *namespace_id = "Namespace identifier";
+	const char *data = "optional file for data (default stdin)";
+	const char *data_len = "length of data in bytes";
+
+	struct nvme_dev *dev;
+	void *buf = NULL;
+	int err = -1;
+	int dfd = STDIN_FILENO;
+
+	struct config {
+		__u16 mos;
+		__u8  mo;
+		__u32 namespace_id;
+		char  *file;
+		__u32 data_len;
+	};
+
+	struct config cfg = {
+		.mos = 0,
+	};
+
+	OPT_ARGS(opts) = {
+		OPT_UINT("namespace-id",  'n', &cfg.namespace_id,   namespace_id),
+		OPT_UINT("mos",           's', &cfg.mos,            mos),
+		OPT_BYTE("mo",            'm', &cfg.mo,             mo),
+		OPT_FILE("data",          'd', &cfg.file,           data),
+		OPT_UINT("data-len",      'l', &cfg.data_len,       data_len),
+		OPT_END()
+	};
+
+	err = parse_and_open(&dev, argc, argv, desc, opts);
+	if (err)
+		return errno;
+
+	if (!cfg.namespace_id) {
+		err = nvme_get_nsid(dev_fd(dev), &cfg.namespace_id);
+		if (err < 0) {
+			perror("get-namespace-id");
+			goto close_dev;
+		}
+	}
+
+	if (cfg.data_len) {
+		buf = calloc(1, cfg.data_len);
+		if (!buf) {
+			perror("could not alloc memory for io mgmt receive data");
+			err = -ENOMEM;
+			goto close_dev;
+		}
+	}
+
+	if (cfg.file) {
+		dfd = open(cfg.file, O_RDONLY);
+		if (dfd < 0) {
+			perror(cfg.file);
+			goto free;
+		}
+	}
+
+	err = read(dfd, buf, cfg.data_len);
+	if (err < 0) {
+		perror("read");
+		goto close_fd;
+	}
+
+	struct nvme_io_mgmt_send_args args = {
+		.args_size	= sizeof(args),
+		.fd		= dev_fd(dev),
+		.nsid		= cfg.namespace_id,
+		.mos		= cfg.mos,
+		.mo		= cfg.mo,
+		.data_len	= cfg.data_len,
+		.data		= buf,
+		.timeout	= NVME_DEFAULT_IOCTL_TIMEOUT,
+	};
+
+	err = nvme_io_mgmt_send(&args);
+	if (!err)
+		printf("io-mgmt-send: Success, mos:%u mo:%u nsid:%d\n",
+			cfg.mos, cfg.mo, cfg.namespace_id);
+	else if (err > 0)
+		nvme_show_status(err);
+	else
+		perror("io-mgmt-recv");
+
+close_fd:
+	if (cfg.file)
+		close(dfd);
+free:
+	free(buf);
+close_dev:
+	dev_close(dev);
+	return err;
+}
+
+static int io_mgmt_recv(int argc, char **argv, struct command *cmd, struct plugin *plugin)
+{
+	const char *desc = "I/O Management Receive";
+	const char *mos = "Management Operation Specific";
+	const char *mo = "Management Operation";
```

Comment:
@ikegami-t  was busy grouping common strings at the beginning of the file. I suppose we should these also there.

Reply:
Okidokie.